Remove old code examples and add tagged-events instructions

RobertJoonas opened the pull request:
This PR is to be merged once the `tagged-events` script extension is live. Pending PR here: https://github.com/plausible/analytics/pull/2333

Read formatted markdown here: https://github.com/plausible/docs/blob/9e605dd475f89af6e2d8a8f73462691e95201b82/docs/custom-event-goals.md (up to date as of Nov 22, 07:16 UTC)

metmarkosaric commented:
thanks @RobertJoonas! shouldn't the new tagged-events instructions just replace what we have on https://plausible.io/docs/custom-event-goals rather than adding a completely new page about the same topic? we can keep the manual method there too for those that prefer it

RobertJoonas commented:
Thanks for the feedback, @metmarkosaric! 

> shouldn't the new tagged-events instructions just replace what we have on https://plausible.io/docs/custom-event-goals rather than adding a completely new page about the same topic?

I wouldn't say it's on the exact same topic. True, its goal is to set up custom events, but it's more like a separate page with instructions for using one specific feature. Both `outbound-links` and `file-downloads` are also for tracking custom events, and they have their own doc pages. Also, it's a script extension that we can link to from the [script-extensions page](https://plausible.io/docs/script-extensions).

There are specifics that only apply for the `tagged-events` feature, and not custom events in general:

- adding class names to HTML elements
- tracking behaviors on different elements.

We could put it all in the custom-event-goals page, but IMO it would look a bit bloated. I couldn't come up with a readable structure like that, so I decided to go with this section in the custom-event-goals doc:

![image](https://user-images.githubusercontent.com/56999674/198995367-c7b62341-c600-4fec-8fb6-34f51be96549.png)

That way, everything we support out-of-the-box is separated from everything else that requires putting custom JavaScript code on your site (+ the code examples).

However, it does look a bit hard to find these instructions at the moment. What do you think about moving the "Out-of-the-box custom event integrations" section to the top of the custom-event-goals document? 

> we can keep the manual method there too for those that prefer it

What do you mean by "manual method" here?

metmarkosaric commented:
ok. i'm thinking about this from the user's perspective and our old and our new custom event instructions basically describe solutions to the same problem so having that on two separate pages is what i prefer to avoid. it creates confusion plus makes it more difficult to discover.

technically these are different things in the way we've built them but they're just slightly different ways to achieve the same goal for the user. they don't really care how it works in our background, they care what they want to do (track signups or purchases etc). our new method is the most user-friendly one and the easiest one so should become the norm that we present to everyone in the future. we can still keep the [current step 1](https://plausible.io/docs/custom-event-goals#1-trigger-custom-events-with-javascript-on-your-site) and say something like "if you prefer to do this manually, here what you can do".

keeping these on the same page doesn't prevent us from including links to it on the extension page etc.

outbound link clicks and file downloads are completely different features for people even though they're still the same custom events technically so they deserve to be separate.

i don't know how to combine these in the pull request itself but if you can combine it onto the main custom-event-goals.md page, then i can go through it myself and edit it to make it even simpler than now. thanks!

RobertJoonas commented:
Thanks, @metmarkosaric, good points all around! That makes total sense from the users' perspective. I've now replaced the commit I made before with a new one.

I deleted the extra page and included it all on the custom-event-goals page. Also moved the headings around a little bit so that it would make sense.

One thing though, we should mention adding the goal from the dashboard UI also for the "do it manually" part (second level-1 heading). Perhaps just include the same section twice for both tagged events and manual?

...Also dropped in a link to the script extensions page.

metmarkosaric commented:
nice, thanks @RobertJoonas! i wonder if markdown or docusaurus can do "click to display" kind of section? (wordpress allows that so i know it from there). so we have sat like "if you prefer to do this manually, click to expand instructions for step 1" and then it just expands the page with that info. that would make the page cleaner / shorter but still allow us to display that info for those who want

RobertJoonas commented:
> i wonder if markdown or docusaurus can do "click to display/expand" kind of section?

This it's possible by wrapping the whole thing in a `<details>` element, directly in markdown:

```
<details>
<summary>

## heading to be clicked for expanding

</summary>

## all other content to be shown

</details>
```

https://user-images.githubusercontent.com/56999674/199117307-cd28b6ef-d4da-42b2-a239-cc2da6bc4a2c.mp4

I think it's also possible to override the background color and border, but IMO this default light-blue looks good. What do you think @metmarkosaric?

metmarkosaric commented:
that looks great to me @RobertJoonas!

ukutaht commented on docs/custom-event-goals.md:
```diff
@@ -67,7 +137,54 @@ Select `Custom event` as the goal trigger and enter the name of the custom event
+- Triggers events when the link is clicked/middle-clicked with a mouse, or touched on mobile.
+- A property called `url` is automatically added as a custom event property with the `href` value of the link (that is the URL where the link points to.
+
+#### Form - `<form>`
```
I don't want to complicate this piece of work too much but I wanted to have one last discussion about an idea I had earlier on tagged events.

Specially with forms, it might be useful to track more things than just submit. Once we implement funnels, it might be really nice to also track the number of people who **interact** with the form. Interaction would mean they started filling out at least one of the form inputs.

You can search the web for 'Form analytics'. It's a whole industry. With proper form analytics you can build funnels like:

```
* Visited register page - 300 visitors
* Interacted with Signup form - 100 visitors
* Submitted Signup form - 50 visitors
```

This would allow you to see that 50 visitors abandoned the form after starting to fill it out.

Form dropoff rates are one of the most useful pieces of information one can get from their analytics because it informs how difficult it is to fill out a form. In my experience, anyone working on conversion optimization is interested in this data. Advanced form analytics tools also tell you which form fields visitors get most stuck with.

After writing this out, and especially with 'form analytics' being a whole industry with it's own vocabulary and standard practices, I think the best approach is to leave this kind of more complicated form analysis out of this script and consider creating a new script extension for 'Form Analytics' in the future if people ask for it.

ukutaht commented on docs/custom-event-goals.md:
```diff
@@ -111,117 +228,30 @@ Note that you must ensure that no personally identifiable information (PII) is s
-:::note
-Do watch the quotes! The `data-analytics` tag value should have both **single and double quotes** as shown above.
-:::
+### Tracking form submits
```
This seems redundant now. We can start using tagged events internally as well and remove this from the docs.

RobertJoonas replied:
Done, removed in the latest commit.

ukutaht commented on docs/custom-event-goals.md:
````diff
@@ -111,117 +228,30 @@ Note that you must ensure that no personally identifiable information (PII) is s
 ```
 
-## Contact Form 7 plugin on WordPress
+### Contact Form 7 plugin on WordPress
````
Can we simplify how this works using tagged-events as well?

Ideally our WP plugin would just recognise that it's running alongside Contact Form 7 and allow people a simple one-click integration in the future.

RobertJoonas replied:
I feel like it wouldn't fit into this feature. Sounds more like an edge case that has very straightforward instructions as it is.

But perhaps we can still build it into the WP plugin as a one-click integration, just adding that script to the site where the form is.

ukutaht reviewed (approved):
I had some small wording suggestions but this looks very clear and easy to use now. Thanks @RobertJoonas !

metmarkosaric commented:
"form analytics" sounds great to me!

i'll remove the contact form 7 part as it's not necessary. i didn't actually make any changes myself yet as it's difficult for me to see on github how it all looks in the actual browser. so when we're ready with this i can merge, take a look at the page in the browser and make some final changes too

RobertJoonas commented on docs/custom-event-goals.md:
```diff
@@ -8,52 +8,122 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
 
-  plausible('Signup', {callback: submitForm});
-})
+You can use this `id` to add the class names to your element dynamically with JavaScript. Here's the code you will have to insert in the `<head>` section of the page with the element you want to track:
```
> I'm not sure how many people need something like this.

This enables tagged-events for Webflow, Squarespace, and probably many other CMSs that don't allow custom CSS classes.

> But if we want to make it even easier we could build it into the tagged-events script itself. The API could have roughly this shape

Nice idea, this would look cleaner indeed. I would personally go with this approach, too. Just some opposing thoughts:

1) A user would still have to insert that script on their site, so the integration complexity would roughly stay the same
2) Every user would have this functionality, even though most would probably not need it

RobertJoonas commented on docs/custom-event-goals.md:
```diff
@@ -8,52 +8,122 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
+
+## 2. Add special classes to the elements you want to track
+
+In order to hook up a custom event to your chosen element, you will need to tag it with some special class names. In addition to telling Plausible that the element should be tracked, these class names also determine the name and properties of your custom event.
```
👍 Done in https://github.com/plausible/docs/pull/295/commits/9e605dd475f89af6e2d8a8f73462691e95201b82

RobertJoonas commented on docs/custom-event-goals.md:
```diff
@@ -8,52 +8,122 @@ import useBaseUrl from '@docusaurus/useBaseUrl';
+In some cases, the tracking classes might be added to a wrapper `<div>` element (parent to the element you want to track), but don't worry, Plausible will still be able to track clicks on the child element if its parent has the necessary classes. 
 
-If you want to trigger form submissions, button clicks etc. you need to implement the code for that yourself. This is an example of code we use to track custom events (trial signups) on our own website. We have listed few other code examples further down this page.
+If the class name format has been changed by your CMS, Plausible will most likely not be able to track it. If that's the case, please keep reading the below instructions.
```
👍 Done in https://github.com/plausible/docs/pull/295/commits/9e605dd475f89af6e2d8a8f73462691e95201b82

RobertJoonas commented:
> i'll remove the contact form 7 part as it's not necessary. i didn't actually make any changes myself yet as it's difficult for me to see on github how it all looks in the actual browser. so when we're ready with this i can merge, take a look at the page in the browser and make some final changes too

@marko, I have now removed the contact form 7 part myself in the latest commit. (Formatted preview is up to date again)

I think there's no way to gradually release this documentation because it will replace the existing custom events document... Should we just share the same [formatted markdown preview document](https://github.com/plausible/docs/blob/9e605dd475f89af6e2d8a8f73462691e95201b82/docs/custom-event-goals.md)?

metmarkosaric commented:
sure @RobertJoonas, that will work for testing purposes
